File: go/day7/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	input, _ := os.ReadFile("go/day7/input.txt")
	root := parseDirectories(string(input))
	var sizes []uint64
	sizes, _ = root.allDirSizes(sizes)
	var smallest uint64 = 9999999999
	var totalSizeUnder100k uint64
	const (
		totalMemory    = 70000000
		requiredMemory = 30000000
	)
	usedMemory := root.size()
	freeMemory := totalMemory - usedMemory
	moreSpaceNeeded := requiredMemory - freeMemory
	for _, s := range sizes {
		if s < 100_000 {
			totalSizeUnder100k += s
		}
		if s > moreSpaceNeeded && s < smallest {
			smallest = s
		}
	}
	fmt.Printf("Part 1. The sum of all directory sizes under 100k is %d\n", totalSizeUnder100k)
	fmt.Printf("Part 2. The smallest directory you could delete to get the required space is %d\n", smallest)
}

func parseDirectories(input string) directory {
	root := directory{name: "/"}
	currentDirectory := &root
	previousDirectories := []directory{}
	for _, line := range strings.Split(input, "\r\n") {
		if strings.HasPrefix(line, "$ cd") {
			target := line[len("$ cd "):]
			switch target {
			case "/":
				currentDirectory = &root
				previousDirectories = []directory{}
			case "..":
				currentDirectory = &previousDirectories[len(previousDirectories)-1]
				// Pop last previous directory.
				if len(previousDirectories) > 1 {
					previousDirectories = previousDirectories[:len(previousDirectories)-1]
				} else {
					previousDirectories = []directory{}
				}
			default:
				var subDirectory *directory
				for _, sub := range currentDirectory.subdirectories {
					if sub.name == target {
						subDirectory = sub
						break
					}
				}
				previousDirectories = append(previousDirectories, *currentDirectory)
				currentDirectory = subDirectory
			}
		} else if line[0] != '$' {
			if strings.HasPrefix(line, "dir") {
				dir := &directory{
					name: line[len("dir "):],
				}
				currentDirectory.subdirectories = append(currentDirectory.subdirectories, dir)
			} else { // File.
				words := strings.SplitN(line, " ", 2)
				size, _ := strconv.ParseUint(words[0], 10, 32)
				file := file{
					name: words[1],
					size: size,
				}
				currentDirectory.files = append(currentDirectory.files, &file)
			}
		}
	}
	return root
}

type directory struct {
	name           string
	subdirectories []*directory
	files          []*file
}

type file struct {
	name string
	size uint64
}

func (d directory) size() uint64 {
	var sum uint64
	for _, f := range d.files {
		sum += f.size
	}
	for _, sub := range d.subdirectories {
		sum += sub.size()
	}

	return sum
}

func (d directory) allDirSizes(sizes []uint64) ([]uint64, uint64) {
	var sum uint64
	for _, f := range d.files {
		sum += f.size
	}
	for _, sub := range d.subdirectories {
		var subSum uint64
		sizes, subSum = sub.allDirSizes(sizes)
		sum += subSum
	}

	sizes = append(sizes, sum)
	return sizes, sum
}

File: src/day7/main.rs
use std::fs;

// I got too annoyed and switched to Go this day. I'll commit this anyway...

fn main() {
    let input = fs::read_to_string("src/day7/input.txt").expect("Should be able to read the input");
    parse_directories(&input);
}

fn parse_directories(input: &str) -> Directory {
    let mut root = Directory::new(String::from("/"));
    // let mut current_directory = &mut root;
    let mut directories: Vec<&mut Directory> = vec![&mut root];
    for line in input.lines() {
        // if line.starts_with("$ ls") {
        // continue;
        if line.starts_with("$ cd") {
            let target = line.strip_prefix("$ cd ").unwrap();
            if target == ".." {
                directories.pop();
            } else {
                // directories.push(current_directory);
                directories.push(
                    directories
                        .last_mut()
                        .unwrap()
                        .subdirectories
                        .iter_mut()
                        .find(|d| d.name == target)
                        .unwrap(),
                )
            }
        } else if !line.starts_with('$') {
            if line.starts_with("dir") {
                directories
                    .last_mut()
                    .unwrap()
                    .subdirectories
                    .push(Directory::new(line.split_once(' ').unwrap().1.to_string()));
            } else {
                let (size, name) = line.split_once(' ').unwrap();
                directories
                    .last_mut()
                    .unwrap()
                    .files
                    .push(File::new(name.to_string(), size.parse().unwrap()))
            }
        }
    }

    root
}

struct Directory {
    name: String,
    subdirectories: Vec<Directory>,
    files: Vec<File>,
}

struct File {
    name: String,
    size: u64,
}

impl Directory {
    fn new(name: String) -> Directory {
        Directory {
            name,
            subdirectories: vec![],
            files: vec![],
        }
    }

    fn size(&self) -> u64 {
        self.files.iter().map(|file| file.size).sum()
    }
}

impl File {
    fn new(name: String, size: u64) -> File {
        File { name, size }
    }
}
